src/lib.rs

```diff
@@ -60,98 +41,251 @@ fn set_stack_limit(l: usize) {
-                fn get_tib_address() -> *const usize;
+        extern {
+            fn __stacker_get_current_fiber() -> winapi::PVOID;
+        }
+
+        #[no_mangle]
+        pub unsafe extern fn __stacker_switch_stacks_callback(f: &mut &mut FnMut()) {
+            f();
+        }
+
+        struct FiberInfo<'a> {
+            callback: &'a mut FnMut(),
+            result: Option<std::thread::Result<()>>,
+            parent_fiber: winapi::LPVOID,
+        }
+
+        unsafe extern "system" fn fiber_proc(info: winapi::LPVOID) {
+            let info = &mut *(info as *mut FiberInfo);
+            info.result = Some(std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+                (info.callback)();
+            })));
+            kernel32::SwitchToFiber(info.parent_fiber);
+            return;
+        }
+
+        fn _grow(stack_size: usize, callback: &mut FnMut()) {
+            unsafe {
+                let was_fiber = kernel32::IsThreadAFiber() == winapi::TRUE;
+                
+                let mut info = FiberInfo {
+                    callback,
+                    result: None,
+                    parent_fiber: if was_fiber {
+                        __stacker_get_current_fiber()
+                    } else {
+                        kernel32::ConvertThreadToFiber(0i32 as _)
+                    },
+                };
+                if info.parent_fiber == 0i32 as _ {
+                    panic!("unable to convert thread to fiber");
+                }
+                let fiber = kernel32::CreateFiber(stack_size as _, Some(fiber_proc), &mut info as *mut FiberInfo as *mut _);
+                if fiber == 0i32 as _ {
+                    panic!("unable to allocate fiber");
+                }
+                kernel32::SwitchToFiber(fiber);
+                kernel32::DeleteFiber(fiber);
+
+                if !was_fiber {
+                    kernel32::ConvertFiberToThread();
+                }
+
+                if let Err(payload) = info.result.unwrap() {
+                    std::panic::resume_unwind(payload);
+                }
             }
-            #[cfg(target_pointer_width = "64")]
-            extern "system" {
-                #[cfg_attr(target_env = "msvc", link_name = "NtCurrentTeb")]
-                #[cfg_attr(target_env = "gnu", link_name = "__stacker_get_tib_64")]
-                fn get_tib_address() -> *const usize;
+        }
+
+        cfg_if! {
+            if #[cfg(any(target_arch = "x86_64", target_arch = "x86"))] {
+                extern {
+                    fn __stacker_get_stack_limit() -> usize;
+                }
+
+                #[inline(always)]
+                fn get_stack_limit() -> usize {
+                    unsafe {
+                        __stacker_get_stack_limit()
+                    }
+                }
+            } else {
+                #[inline(always)]
+                fn get_thread_stack_guarantee() -> usize {
```

Comment:
Is this used because aren't all windows platforms x86/x86_64 righ tnow basically?

Reply:
I did see someone trying to use Rust on ARM64